neutrons/SNAPRed
# Ewm3715 iterable workflows (DiffractionCalibration)

## Description of work



This enables:
1. workflows to be performed iteratively
2. easy, generic, hard and soft widget resets
3. skippable workflow steps
4. Clearing ads, excluding cache and specified workspaces.
5. exposing rename and clear options for workspaces in the ui. 

Refactored:
1. DiffractionCalibrationWorkflow
  i. added a WorflowImplementer class with standardized members and utilities that should be shard across implemented workflows 

## Explanation of work

Resettable was implemented as a decorator because they are easy to adopt and do no pollute the lineage of a class hierarchy while still providing the utilities a widget needs.  It was also necessary to some degree, as it seemed most straight forward to intercept init args in order to generically reset and self contained widget.


## To test

GUI testing


NOTE: Diamond might be broken at the moment, please try other test configs if so
```
On analysis.sns.gov
From SNAPRed directory and environment
python -m snapred
Open "Test Panel" and enter:
Run Number: 46680
Convergence Threshold: 0.1
Peak Intensity Threshold: 0.0001
Bins Across Peak Width: 10
Sample: Diamond_001.json
Grouping File: column lite
Click Continue
Click Continue on Assessment Step as well
Click Skip on Save Step
Click Continue on Go Again? Step
Repeat
Observe output Workpaces are retained between iterations..
```


## Link to EWM item


[EWM#3715](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=3715)
